### ISSUE TITLE
When using MagicMode on allready created property the setter is called twice

### ISSUE THREAD

**Issue by petterek** [ractivejs/ractive#1639]
When using Object.defineProperty in viewmodel, setting the value seems to 
be triggerd twice. 
The value is bound to a dropdown

    <select value="{{selectedFiletype}}">
        {{#fileTypes}}
        <option value="{{.}}">{{Name}}</option>
        {{/f}}
    </select>

Code that is called

    iRepository.ViewDocumentTypesForUser({ UnitId: data.Id }).then(
        function (res) {
             self.fileTypes = res;
        });
    

    var _selectedFiletype;

    Object.defineProperty(this, "selectedFiletype", {
        configurable: true,
        get: function () {
            return _selectedFiletype;
        },
        set: function (value) {
            _selectedFiletype = value;
            debugger;
            iRepository.GetDocumentListForUnitQuery({ UnitId: self.User.Id, 
    DocumentTypeId: value.Id }).then(
                function (res) {
                    self.allItems = res;
            });

        }
    });

Then the setter for selectedFileType is called twice when I change the 
dropdown

**Comment by Madgvox** [comment]
Would it be possible for you to use a computed value for `selectedFiletype`? It behaves just like `Object.defineProperty`, and it doesn't muck with Ractive's own listeners.

```js
computed: {
  selectedFiletype: {
    get: function () {
        return _selectedFiletype;
    },
    set: function (value) {
        _selectedFiletype = value;
        debugger;
        iRepository.GetDocumentListForUnitQuery({ UnitId: self.User.Id, DocumentTypeId: value.Id }).then(
            function (res) {
                self.allItems = res;
        });

    }
  }
}
```

**Comment by petterek** [comment]
No I wil not be able to use the "computed" feature. 
But is this a bug? Or am I doing something wrong?

**Comment by Madgvox** [comment]
I don't know. I'd need to see more context or an isolated test case since I'm not even sure what the 'this' in your snippet is pointing at.

**Comment by petterek** [comment]
Ok, I will try to make a fiddel to reproduce it

**Comment by petterek** [comment]
http://jsfiddle.net/yxyh3omn/

Seems same thing is happening here

**Comment by petterek** [comment]
Except when changing to the 'last' value of the list

**Comment by petterek** [comment]
Acually it seems to call the sette 2 times every second time I change the dropdown

**Comment by Madgvox** [comment]
Ok, so I stepped through the code and this is essentially what's happening.

What happens when the setter is only called once:

1. Select a result from the dropdown. This triggers `Ractive.set`.
2. `Ractive.set` internally triggers `Ractive.get`, in order to sync up the model.
3. Since `magic: true`, Ractive wraps the returned value in a `MagicMode` wrapper that returns a 'magic' object.
4. That `MagicMode` wrapper overwrote your `set` method, so in its own `set` it calls your `set` manually.
5. __`SelectCount++`.__

What happens when the setter is called twice:

1. Select a result from the dropdown. This triggers `Ractive.set`.
2. `Ractive.set` sees that this value is wrapped and 'resets' the wrapper to the new value.
3. The wrapper updates the value on the object, which triggers the object's wrapped `set`, which triggers your `set`.
4. __`SelectCount++`.__
5. Since the value has changed, Ractive clears its cache.
6. Since the value is currently wrapped with the `MagicMode` wrapper, it needs to teardown that wrapper in order to safely change the value.
7. The `teardown` function cleans up its custom `get` and `set` methods and restores your original `get` and `set` onto the object.
8. Then it sets the object's value back to what it's supposed to be.
9. This triggers your `set`.
10. __`SelectCount++`.__

So basically the MagicMode wrapper doesn't deal with objects with getters and setters properly. It probably shouldn't do step 8 if the object had getter and setter listeners originally.

**Comment by petterek** [comment]
Is this something that will be fixed?
I solved it in my code by 

    if (_selectedFiletype == value) return;

Thank you for investegating

**Comment by Madgvox** [comment]
No problem, but I wouldn't close this yet. This is a bug with magic mode that needs to be addressed.

**Comment by petterek** [comment]
OK.

**Comment by petterek** [comment]
Did the fix above solve the issue I filed?

**Comment by Madgvox** [comment]
No, but I did make a test for this issue. It took me so long and it is so specific that I don't think it is worth it to fix. If it pops up again we'll take another look, but right now it's just too much of an edge case.
